main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,242 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur.*
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.{DefContext, Occur}
+import ca.uwaterloo.flix.language.ast.Symbol.{DefnSym, VarSym}
+import ca.uwaterloo.flix.language.ast.{OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.ParOps
+
+/**
+  * The occurrence analyzer collects information on variable and function usage and calculates the weight of the expressions
+  * Marks a variable or function as Dead if it is not used, Once if it is used exactly once and Many otherwise
+  */
+object OccurrenceAnalyzer {
+
+  private object OccurInfo {
+
+    /**
+      * Occurrence information for an empty sequence of expressions
+      */
+    val Empty: OccurInfo = OccurInfo(Map.empty, Map.empty, 0, 0)
+
+    /**
+      * The initial occurrence information for an expression of size 1, i.e. an expression without subexpressions.
+      */
+    val One: OccurInfo = OccurInfo(Map.empty, Map.empty, 0, 1)
+  }
+
+  /**
+    * The occurrence of `defs` and `vars` inside the body of a `def`
+    * `size` represents the number of expressions in the body of a `def`
+    */
+  case class OccurInfo(defs: Map[DefnSym, Occur], vars: Map[VarSym, Occur], localDefs: Int, size: Int) {
+    def :+(kv: (DefnSym, Occur)): OccurInfo = {
+      this.copy(defs = this.defs + kv)
+    }
+
+    def +(kv: (VarSym, Occur)): OccurInfo = {
+      this.copy(vars = this.vars + kv)
+    }
+
+    def -(varSym: VarSym): OccurInfo = {
+      this.copy(vars = this.vars - varSym)
+    }
+
+    def --(varSyms: Iterable[VarSym]): OccurInfo = {
+      this.copy(vars = this.vars -- varSyms)
+    }
+
+    def get(defnSym: DefnSym): Occur = {
+      this.defs.getOrElse(defnSym, Dead)
+    }
+
+    def get(varSym: VarSym): Occur = {
+      this.vars.getOrElse(varSym, Dead)
+    }
+
+    def incrementLocalDefCount: OccurInfo = {
+      this.copy(localDefs = localDefs + 1)
+    }
+  }
+
+  private def increment(occurInfo: OccurInfo): OccurInfo = {
+    occurInfo.copy(size = occurInfo.size + 1)
+  }
+
+  /**
+    * A set of functions that contain masked casts.
+    *
+    * Must be manually maintained since Lowering erases the masked cast.
+    */
+  private val DangerousFunctions: Set[String] = Set("bug!", "Fixpoint.Debugging.notifyPreSolve", "Fixpoint.Debugging.notifyPostSolve", "Fixpoint.Debugging.notifyPreInterpret", "Assert.eq")
```

Comment:
Ugh.

What is happening here?

Is this essentially a blacklist of functions not to touch?

Reply:
Yes exactly

Reply:
This is the long discussion we had on inlining holes and casts. Since it is much harder to figure out where holes and casts are and whether they are safe to inline, we never inline functions that contain these and explicitly ban functions that contain masked casts (because they lie even more to the compiler).

Reply:
Ah ok check this out https://github.com/flix/flix/pull/10335#discussion_r2054170912 -- masked casts are gone!